### PR TITLE
feat(FN-3267): update facility utilisation data with keying data

### PR DESCRIPTION
**Pull request by JoshBinns2000**
## Introduction :pencil2:
We want to update the facility utilisation data table when keying data is generated for a facility.

### Note
Due to the fact that keying data generation modifies data on the actual objects we pass in, it is important that the updates are performed in the correct order ie. the principal balance adjustment and fixed fee must be calculated before the facility utilisation data is updated. To ensure this is done, a check is performed in the `calculateFixedFeeAdjustment` function which ensures that the facility utilisation data report period is not the same as the current report report period which should be sufficient as a check for whether or not the facility utilisation data has been updated prematurely.

## Resolution :heavy_check_mark:
- Adds a new helper in the event handler to update the facility utilisation data
- Updates unit and api tests

### CODE REVIEW

File: libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts [hunk 1]
```diff
@@ -42,4 +48,11 @@ export class FacilityUtilisationDataEntity extends AuditableBaseEntity {
     data.updateLastUpdatedBy(requestSource);
     return data;
   }
+
+  public updateWithFixedFeeUtilisationAndReportPeriod({ fixedFee, utilisation, reportPeriod, requestSource }: UpdateWithFixedFeeAndReportPeriodParams): void {
```
Comment (sw-mattw) [inline libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts:52]: Bit of a nitpick on the naming here:

Utilisation and fixed fee are separated above (`createWithoutUtilisationAndFixedFee`), but joined here (`updateWithFixedFeeUtilisationAndReportPeriod`, the `FixedFeeUtilisation` substring) and utilisation is then excluded in the type `UpdateWithFixedFeeAndReportPeriodParams`.

Could be good to make these all the same, eg:
- createWithoutUtilisationAndFixedFee -> createWithoutFixedFeeAndUtilisation
- updateWithFixedFeeUtilisationAndReportPeriod -> updateWithFixedFeeAndUtilisationAndReportPeriod
- UpdateWithFixedFeeAndReportPeriodParams -> UpdateWithFixedFeeAndUtilisationAndReportPeriodParams

File: libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.types.ts [hunk 2]
```diff
@@ -6,3 +6,10 @@ export type CreateFacilityUtilisationDataWithoutUtilisationParams = {
   reportPeriod: ReportPeriod;
   requestSource: DbRequestSource;
 };
+
+export type UpdateWithFixedFeeAndReportPeriodParams = {
```
Comment (sw-mattw) [inline libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.types.ts:10]: Nitpick - see prior comment on naming of this on `libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts`

**Review by sw-mattw — CHANGES_REQUESTED** [review]
Looks good, just some minor typo and tidy comments and a query about splitting a big test up

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/post-keying-data.controller/index.test.ts [hunk 3]
```diff
@@ -78,8 +78,13 @@ describe('post-keying-data.controller', () => {
...
       const { req, res } = getHttpMocks();
 
-      const undefinedReport = undefined as unknown as UtilisationReportEntity;
-      when(feeRecordRepoFindSpy).calledWith(reportId, ['MATCH']).mockResolvedValue(someFeeRecordsForReport(undefinedReport));
+      const feeRecords = someFeeRecordsForReport(UtilisationReportEntityMockBuilder.forStatus('RECONCILIATION_IN_PROGRESS').build());
+      feeRecords.forEach((feeRecord) => {
+        // @ts-expect-error We are setting the report to be undefined purposefully
+        // eslint-disable-next-line no-param-reassign
+        delete feeRecord.report;
+      });
```
[the remark this answers is not in the document]
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/post-keying-data.controller/index.test.ts:86]: The problem with trying to do that is that `feeRecord` is a class and destructing a class loses the prototype chain (ie. we lose all the methods and such)

[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/post-keying-data.controller/index.test.ts:86]: Makes sense to keep it as it is then, thanks for clarifying

File: libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts (hunk 1 above)
Comment (JoshBinns2000) [reply libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts:52]: This naming convention is based on the naming convention we tend to follow in test files ie. `withABAndC` should read as `with A, B and C`. However, the naming convention we have tended to follow in other entities has been to name the `update` method with something context specific. In this case, I think a good choice for the name will be `updateWithCurrentReportPeriodDetails`

File: dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts [hunk 4]
```diff
@@ -375,5 +366,83 @@ describe(`POST ${BASE_URL}`, () => {
       expect(feeRecordWithKeyingData).toHaveLength(1);
       expect(feeRecordWithKeyingData[0].id).toBe(1);
     });
+
+    it('only updates the facility utilisation data table when all the fee records are at the MATCH status', async () => {
```
[the remark this answers is not in the document]
Comment (JoshBinns2000) [reply dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts:370]: I've created a new describe block setup inside this block for both the tests which have two arrange/act/assert sets. Let me know what you think

File: libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts (hunk 1 above)
Comment (sw-mattw) [reply libs/common/src/sql-db-entities/facility-utilisation-data/facility-utilisation-data.entity.ts]: That's great, thank you for following up

File: dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (sw-mattw) [reply dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts]: Looks great thanks for actioning

**Review by sw-mattw — APPROVED** [review]
Looks good, thanks for the changes

**Pull request merged**